addons/account/migrations/13.0.1.1/post-migration.py

```diff
@@ -0,0 +1,781 @@
+        create_uid, create_date, write_uid, write_date)
+        SELECT ait.company_id, ait.account_id, ait.sequence, ait.name,
+        ait.amount, ait.currency_id, ait.base, ait.tax_id,
+        ait.account_analytic_id, COALESCE(ai.move_id, am.id),
+        ait.id, ait.create_uid, ait.create_date, ait.write_uid, ait.write_date
+        FROM account_invoice_tax ait
+        JOIN account_invoice ai ON ait.invoice_id = ai.id AND ai.state IN ('draft', 'cancel')
+        LEFT JOIN account_move am ON am.old_invoice_id = ai.id
+        WHERE COALESCE(ai.move_id, am.id) IS NOT NULL""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE account_invoice_payment_rel aipr
+        SET invoice_id = am.id
+        FROM account_move am
+        WHERE am.old_invoice_id = aipr.invoice_id
+        """,
+    )
+    # Relink analytic tags
+    openupgrade.logged_query(
+        env.cr, """
+        INSERT INTO account_analytic_tag_account_move_line_rel (
+            account_move_line_id, account_analytic_tag_id)
+        SELECT aml.id, aatailr.account_analytic_tag_id
+        FROM account_analytic_tag_account_invoice_line_rel aatailr
+        JOIN account_move_line aml
+        ON aml.old_invoice_line_id = aatailr.account_invoice_line_id
+        LEFT JOIN account_analytic_tag_account_move_line_rel aatamlr
+        ON aml.id = aatamlr.account_move_line_id AND aatailr.account_analytic_tag_id = aatamlr.account_analytic_tag_id
+        WHERE aatamlr.account_analytic_tag_id IS NULL
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        INSERT INTO account_analytic_tag_account_move_line_rel (
+            account_move_line_id, account_analytic_tag_id)
+        SELECT aml.id, aataitr.account_analytic_tag_id
+        FROM account_analytic_tag_account_invoice_tax_rel aataitr
+        JOIN account_move_line aml
+        ON aml.old_invoice_tax_id = aataitr.account_invoice_tax_id
+        LEFT JOIN account_analytic_tag_account_move_line_rel aatamlr
+        ON aml.id = aatamlr.account_move_line_id AND aataitr.account_analytic_tag_id = aatamlr.account_analytic_tag_id
+        WHERE aatamlr.account_analytic_tag_id IS NULL
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        INSERT INTO account_move_line_account_tax_rel (
+            account_move_line_id, account_tax_id)
+        SELECT aml.id, ailt.tax_id
+        FROM account_invoice_line_tax ailt
+        JOIN account_move_line aml
+        ON aml.old_invoice_line_id = ailt.invoice_line_id
+        LEFT JOIN account_move_line_account_tax_rel amlatr
+        ON aml.id = amlatr.account_move_line_id AND ailt.tax_id = amlatr.account_tax_id
+        WHERE amlatr.account_tax_id IS NULL
+        """,
+    )
+    # Compute balance for Draft Invoice Lines
+    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))])
+    for invoice_line in draft_invoices.line_ids:
```

Comment:
I believe we might have a performance issue here. I have a client with not a massive number of draft vendor bills, say 3,500, and stuck here for 16 hours so far. As it happens, they are all errors anyway so I can just delete and restart but FYI.

Reply:
OK, thanks for the info. I will check for alternatives.

Reply:
@gdgellatly try to use with this changes:
<pre>
    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))])
    for invoice_line in draft_invoices.line_ids:
        vals = invoice_line._get_fields_onchange_subtotal()
        invoice_line.write(vals)
    draft_invoices.with_context(check_move_validity=False)._recompute_dynamic_lines()
</pre>
<pre>
    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))])
    draft_invoices.read([
        'price_unit', 'quantity', 'discount', 'currency_id', 'product_id',
        'partner_id', 'tax_ids', 'move_id', 'price_subtotal',
    ])
    draft_invoices.line_ids._onchange_price_subtotal()
    draft_invoices.with_context(check_move_validity=False)._recompute_dynamic_lines(recompute_all_taxes=True)
</pre>

Reply:
@kos94ok-3D As I said, in this case it was an error in client database, so I deleted them all now anyway.